### PR TITLE
refactor/vim opt

### PR DESCRIPTION
**Pull request by lewis6991**
Some spring-cleaning of `_meta.lua`

- Try to de-abstract the code to make it easier to understand
- Simplify things where possible
- Embed the option metatype (`map`, `set`, etc) into the `info` table which is now lazy populated and remove the `_setup()` function.
- Unify and simplify the metatable logic for `vim.bo` and `vim.wo`.
- Make it so `append`, `prepend` and `remove`:
  - operate on the option in-place and not create a new object
  - Do not return a value

Broke down into lots of small commits for easier reviewing.

### CODE REVIEW

File: runtime/lua/vim/_meta.lua [hunk 1]
```diff
@@ -4,170 +4,112 @@ local vim = assert(vim)
...
-      end
+vim.env = setmetatable({}, {
+  __index = function(_, k)
+    return vim.loop.os_getenv(k)
+  end,
 
-      return a.nvim_get_option_value(k, { buf = bufnr or 0 })
-    end
+  __newindex = function(_, k, v)
+    vim.loop.os_setenv(k, v)
```
Comment (zeertzjq) [inline runtime/lua/vim/_meta.lua:45]: Does this regress #17501?

Comment (lewis6991) [reply runtime/lua/vim/_meta.lua:9]: Potentially, I'll look into it.

**Comment by clason** [comment]
As you are now the one most familiar with the implementation, can you take this opportunity to correct the `lua-options` documentation?

**Comment by craigmac** [comment]
@lewis6991 In breaking changes #14090 could you provide more info, like a before/after usage? It's unclear to me what needs to be done, if anything, on the user's part.

**Comment by clason** [comment]
Nothing needs to be changed, as that was the intended use anyway. Just be aware that if you relied on the return value in some (unintended) way, you need to that differently. (How depends on what you used it for.)

**Comment by lewis6991** [comment]
If for whatever reason you had code like:

```lua
vim.opt.wildignore = vim.opt.wildignore:prepend('*.o')
```

that will no longer work and now should just be:

```lua
vim.opt.wildignore:prepend('*.o')
```

which (as @clason mentioned) was the intended use anyway.

**Pull request merged**